server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -242,74 +298,90 @@ public CachingClusteredClient(
+          // This also causes eager initialization of the **sequences**, aka forking off the direct druid client requests
+          // Sequence result accumulation should still be lazy
+          .collect(Collectors.toList())
+          .stream();
     }
 
-    private Set<ServerToSegment> computeSegmentsToQuery(TimelineLookup<String, ServerSelector> timeline)
+    /**
+     * Create a stream of the partition chunks which are useful in this query
+     *
+     * @param holder The holder of the shard to server component of the timeline
+     *
+     * @return Chunks and the segment descriptors corresponding to the chunk
+     */
+    private Stream<ServerToSegment> extractServerAndSegment(TimelineObjectHolder<String, ServerSelector> holder)
     {
-      final List<TimelineObjectHolder<String, ServerSelector>> serversLookup = toolChest.filterSegments(
-          query,
-          query.getIntervals().stream().flatMap(i -> timeline.lookup(i).stream()).collect(Collectors.toList())
-      );
+      return DimFilterUtils
+          .filterShards(
+              query.getFilter(),
+              holder.getObject(),
+              partitionChunk -> partitionChunk.getObject().getSegment().getShardSpec(),
+              Maps.newHashMap()
+          )
+          .stream()
+          .map(chunk -> new ServerToSegment(
+              chunk.getObject(),
+              new SegmentDescriptor(holder.getInterval(), holder.getVersion(), chunk.getChunkNumber())
+          ));
+    }
 
-      final Set<ServerToSegment> segments = Sets.newLinkedHashSet();
-      final Map<String, Optional<RangeSet<String>>> dimensionRangeCache = Maps.newHashMap();
-      // Filter unneeded chunks based on partition dimension
-      for (TimelineObjectHolder<String, ServerSelector> holder : serversLookup) {
-        final Set<PartitionChunk<ServerSelector>> filteredChunks = DimFilterUtils.filterShards(
-            query.getFilter(),
-            holder.getObject(),
-            partitionChunk -> partitionChunk.getObject().getSegment().getShardSpec(),
-            dimensionRangeCache
-        );
-        for (PartitionChunk<ServerSelector> chunk : filteredChunks) {
-          ServerSelector server = chunk.getObject();
-          final SegmentDescriptor segment = new SegmentDescriptor(
-              holder.getInterval(),
-              holder.getVersion(),
-              chunk.getChunkNumber()
-          );
-          segments.add(new ServerToSegment(server, segment));
-        }
-      }
-      return segments;
+    private Stream<ServerToSegment> computeSegmentsToQuery(TimelineLookup<String, ServerSelector> timeline)
+    {
+      return toolChest
+          .filterSegments(
+              query,
+              query.getIntervals().stream().flatMap(i -> timeline.lookup(i).stream()).collect(Collectors.toList())
+          )
+          .stream()
+          .flatMap(this::extractServerAndSegment)
+          .distinct();
```

Comment:
`Stream` is a very "uncertain" and therefore bad abstraction for any declarations: methods and even local variables. It's even more "uncertain" than `Iterable`, because you don't know if the stream is parallel or not, and you cannot iterate it multiple times. When you operate with a method that returns a Stream, you *have* to go to the declaration of that method and read how that stream was created, to understand what is going on in general. It makes it harder to analyze methods in separation and forces you to keep more information in your head at the same time.

So `Stream` works best only in stream chains that end with terminal operations, and that are wholly observable by one sight because they span several lines of code. But when Stream crosses the borders of methods and classes, it complicates understanding of the code.

I understand using Streams in method return types when you want to avoid extra collection materializations, but not when returning a `Collection` (or even better a `Set`) is executionally the same thing.

Reply:
Also passing Streams across methods makes executional borders to not correspond to declarational borders of the logic (because Streams are lazy), that makes it harder to mentally reason about the execution and performance of the code. I suppose this is a popular practice in hardcore functional Scala programming.